File: src/pratice/binarysearch/랜선자르기1654.java
package pratice.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class 랜선자르기1654 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int K, N;
    public static int[] A;

    public static void main(String[] args) throws IOException {
        input();
        Arrays.sort(A);
        pro();
    }

    private static void pro() {
        long L = 0;
        long R = Integer.MAX_VALUE;
        long res = L;
        while(L <= R){
            long mid = (L + R) / 2;
            if(cutAble(mid)){
                res = mid;
                L = mid + 1;
            }else{
                R = mid - 1;
            }
        }
        System.out.println(res);
    }

    private static boolean cutAble(long mid) {
        long cnt = 0;
        for (int i = 0; i < K ; i++) {
            cnt += (A[i] / mid);
            if(cnt >= N)return true;
        }
        return false;
    }

    private static void input() throws IOException {
        String[] split = br.readLine().split(" ");
        K = Integer.parseInt(split[0]);
        N = Integer.parseInt(split[1]);
        A = new int[K];
        for (int i = 0; i < K; i++) {
            A[i] = Integer.parseInt(br.readLine());
        }
    }
}

File: src/pratice/binarysearch/랜선자르기1654.kt
package pratice.binarysearch

fun main() {
    val (K, N) = readln().split(" ").map { it.toInt() }
    val arr = mutableListOf<Int>()
    repeat(K) {
        arr.add(
            readln().toInt()
        )
    }
    arr.sort()

    lowerBound(arr, N)
}

private fun lowerBound(arr: MutableList<Int>, needLineCnt: Int) {
    var L = 0L
    var R = Int.MAX_VALUE.toLong()
    var res = L
    while (L <= R) {
        val mid = (L + R) / 2
        if (cutAble(arr, mid, needLineCnt)) {
            res = mid
            L = mid + 1
        } else {
            R = mid - 1
        }
    }
    println(res)
}

private fun cutAble(arr: MutableList<Int>, num: Long, needLineCnt: Int): Boolean {
    var cnt = 0L
    for (len in arr) {
        cnt += (len / num).toLong()
    }
    return cnt >= needLineCnt
}
